ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerDelivery.sol

```diff
@@ -0,0 +1,645 @@
+        //  can't be funded
+        if (msg.sender != address(this)) {
+            revert RequesterNotWormholeRelayer();
+        }
+
+        Gas gasLimit = evmInstruction.gasLimit;
+        bool success;
+        {
+            address payable deliveryTarget = payable(fromWormholeFormat(evmInstruction.targetAddress));
+            bytes memory callData = abi.encodeCall(IWormholeReceiver.receiveWormholeMessages, (
+                evmInstruction.payload,
+                evmInstruction.signedVaas,
+                evmInstruction.senderAddress,
+                evmInstruction.sourceChain,
+                evmInstruction.deliveryHash
+            ));
+
+            Gas preGas = Gas.wrap(gasleft());
+
+            // Calls the `receiveWormholeMessages` endpoint on the contract `evmInstruction.targetAddress`
+            // (with the gas limit and value specified in instruction, and `encodedVMs` as the input)
+            (success, targetRevertDataTruncated) = returnLengthBoundedCall(
+                deliveryTarget,
+                callData,
+                gasLimit.unwrap(),
+                evmInstruction.totalReceiverValue.unwrap(),
+                RETURNDATA_TRUNCATION_THRESHOLD
+            );
+
+            Gas postGas = Gas.wrap(gasleft());
+
+            unchecked {
+                gasUsed = (preGas - postGas).min(gasLimit);
+            }
+        }
+
+        if (success) {
+            targetRevertDataTruncated = new bytes(0);
+            status = uint8(DeliveryStatus.SUCCESS);
+
+            ForwardInstruction[] storage forwardInstructions = getForwardInstructions();
+
+            if (forwardInstructions.length > 0) {
+                //Calculate the amount of the transaction fee to refund to the user (multiply the maximum refund by the
+                //  fraction of gas unused)
+                LocalNative transactionFeeRefundAmount = (gasLimit - gasUsed).toWei(
+                    evmInstruction.targetChainRefundPerGasUnused
+                ).asLocalNative();
+                emitForward(gasUsed, transactionFeeRefundAmount, forwardInstructions);
+                status = uint8(DeliveryStatus.FORWARD_REQUEST_SUCCESS);
+            }
+        } else {
+            // Note that forward instructions should always be empty in this case.
+            status = uint8(DeliveryStatus.RECEIVER_FAILURE);
+        }
+    }
+
+    /**
+     * - Checks if enough funds were passed into a forward
+     * - Increases the 'extraReceiverValue' of the first forward in order to use all of the funds
+     * - Publishes the DeliveryInstruction
+     * - Pays the relayer's reward address to deliver the forward
+     *
+     * @param transactionFeeRefundAmount amount of maxTransactionFee that was unused
+     * @param forwardInstructions An array of structs containing information about the user's forward
+     *     request(s)
+     */
+    function emitForward(
```

Comment:
This function is very difficult to follow. More in-code documentation would help, but I think we should consider splitting this function into several functions.